mobile-browser-based-version/server/src/logic/federated/handlers.js

```diff
@@ -0,0 +1,547 @@
+  ) {
+    setTimeout(() => _aggregateWeights(task, round, id), AGGREGATION_COUNTDOWN);
+  }
+}
+
+/**
+ * Request handler called when a client sends a POST request asking for
+ * the averaged model weights stored on server while training a task. The
+ * request is made for a given task and round. The request succeeds once
+ * CLIENTS_THRESHOLD % of clients sent their individual weights to the server
+ * for the given task and round. Every MODEL_SAVE_TIMESTEP rounds into the task,
+ * the requested averaged weights are saved under a JSON file at milestones/.
+ * The request's body must contain:
+ * - the client's ID
+ * - a timestamp corresponding to the time at which the request was made
+ * @param {Request} request received from client
+ * @param {Response} response sent to client
+ */
+export async function aggregationStatus(request, response) {
+  const type = REQUEST_TYPES.AGGREGATION_STATUS;
+
+  const code = checkRequest(request);
+  if (code !== 200) {
+    return failRequest(response, type, code);
+  }
+
+  const task = request.params.task;
+  const round = request.params.round;
+  const id = request.params.id;
+
+  /**
+   * The task was not trained at all.
+   */
+  if (!status.get(task).training && status.get(task).round === 0) {
+    return failRequest(response, type, 403);
+  }
+  /**
+   * No weight was posted for this task's round.
+   */
+  if (!(weightsMap.has(task) && weightsMap.get(task).has(round))) {
+    return failRequest(response, type, 404);
+  }
+
+  logsAppend(request, type);
+  /**
+   * Check whether the requested round has completed.
+   */
+  const aggregated =
+    !status.get(task).training ||
+    (status.get(task).training && round < status.get(task).round);
+  /**
+   * If aggregation occured, make the client wait to get selected again so it can
+   * proceed to other jobs.
+   */
+  if (aggregated) {
+    selectedClients.get(task).delete(id);
+  }
+  response.status(200).send({ aggregated: aggregated });
+}
+
+async function _aggregateWeights(task, round, id) {
+  // TODO: check whether this actually works
+  const serializedAggregatedWeights = await averageWeights(
+    Array.from(weightsMap.get(task).get(round).values())
+  );
+  /**
+   * Save the newly aggregated model to local storage. This is now
+   * the model served to clients for the given task. To save the newly
+   * aggregated weights, here is the (cumbersome) procedure:
+   * 1. create a new TFJS model with the right layers
+   * 2. assign the newly aggregated weights to it
+   * 3. save the model
+   */
+  console.log(`Updating ${task} model`);
+  const modelFilesPath = config.SAVING_SCHEME.concat(
+    path.join(config.MODELS_DIR, task, 'model.json')
+  );
+  const model = await tf.loadLayersModel(modelFilesPath);
+  assignWeightsToModel(model, serializedAggregatedWeights);
+  model.save(path.dirname(modelFilesPath));
+  /**
+   * The round has completed training.
+   */
+  status.get(task).training = false;
```

Comment:
i wouldn't call this training false, but better round completed/pending. (as training is the whole process of FL, going on forever)

Reply:
definitely, I was struggling with the name and ended up choosing a bad one